File: intrinsic_sdk/intrinsic/skills/python/skill_interface_utils.py
# Copyright 2026 Intrinsic Innovation LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Utils for Skill implementations."""

from __future__ import annotations

from intrinsic.resources.proto import resource_handle_pb2
from intrinsic.skills.internal import execute_context_impl

# isort: off

from intrinsic.skills.internal import predict_context_impl
from intrinsic.skills.internal import preview_context_impl

# isort: on
from intrinsic.skills.python import skill_interface


def preview_via_execute(
    skill: skill_interface.Skill[
        skill_interface.TParamsType, skill_interface.TResultType
    ],
    request: skill_interface.PreviewRequest[skill_interface.TParamsType],
    context: skill_interface.PreviewContext,
) -> skill_interface.TResultType:
  """Implements Skill.preview by calling Skill.execute.

  A skill can use this function to implement `preview` by calling
  `preview_via_execute` from within its implementation. E.g.:
  ```
  class MySkill(Skill):
    def preview(self, request: PreviewRequest, context: PreviewContext) -> ...:
      ...
      return preview_via_execute(self, request, context)
  ```

  A skill should only use this util to implement `preview` if its `execute`
  method does not require resources or modify the object world.

  Args:
    skill: The skill instance.
    request: The preview request.
    context: The preview context.

  Returns:
    The response from calling `skill.execute`.
  """
  return skill.execute(
      preview_to_execute_request(request),
      preview_to_execute_context(
          context=context,
          resource_handles={},
      ),
  )



def preview_via_predict(
    skill: skill_interface.Skill[
        skill_interface.TParamsType, skill_interface.TResultType
    ],
    request: skill_interface.PreviewRequest[skill_interface.TParamsType],
    context: skill_interface.PreviewContext,
    result: skill_interface.TResultType,
) -> skill_interface.TResultType:
  """Implements Skill.preview as the most likely outcome of Skill.predict.

  A skill can use this function to implement `preview` by calling
  `preview_via_predict` from within its implementation. E.g.:
  ```
  class MySkill(Skill):
    def preview(self, request: PreviewRequest, context: PreviewContext) -> ...:
      ...
      return preview_via_predict(
          skill=self, request=request, context=context, result=MyResultProto()
      )
  ```

  Args:
    skill: The skill instance.
    request: The preview request.
    context: The preview context.
    result: The result proto into which to unpack the most likely result, or
      None if the skill does not produce a result.

  Returns:
    The unpacked most likely result from calling predict.
  """
  predict_result = skill.predict(
      preview_to_predict_request(request),
      preview_to_predict_context(context),
  )

  if len(predict_result.outcomes) < 1:
    raise ValueError(f"{type(skill).__name__} produced no predictions.")

  most_likely_outcome = max(
      predict_result.outcomes, key=lambda outcome: outcome.probability
  )

  # Record the world updates from the most likely outcome.
  previous_start_time = 0.0
  previous_duration = 0.0
  for idx, expected_state in enumerate(most_likely_outcome.expected_states):
    start_time = (
        (expected_state.start_time.ToNanoseconds() / 1e9)
        if expected_state.HasField("start_time")
        else previous_start_time + previous_duration
    )
    duration = expected_state.time_until_update.ToNanoseconds() / 1e9

    elapsed = start_time - previous_start_time
    if elapsed < 0:
      raise ValueError(
          "Expected state start times must be monotonically increasing"
          f" (state[{idx}] start time: {start_time}, state[{idx - 1}] start"
          f" time: {previous_start_time})."
      )

    for update in expected_state.world_updates.updates:
      context.record_world_update(
          update=update, elapsed=elapsed, duration=duration
      )

      # All updates in this expected state started at the same time, so all but
      # the first update should have `elapsed = 0`.
      elapsed = 0.0

    previous_start_time = start_time
    previous_duration = duration

  # Get the result of the most likely outcome.
  if result is not None:
    most_likely_outcome.result.Unpack(result)

  return result





def preview_to_execute_request(
    request: skill_interface.PreviewRequest[skill_interface.TParamsType],
) -> skill_interface.ExecuteRequest[skill_interface.TParamsType]:
  """Converts a PreviewRequest to an ExecuteRequest."""
  return skill_interface.ExecuteRequest(

      internal_data=request.internal_data,

      params=request.params,
  )


def preview_to_execute_context(
    context: skill_interface.PreviewContext,
    resource_handles: dict[str, resource_handle_pb2.ResourceHandle],
) -> skill_interface.ExecuteContext:
  """Converts a PreviewContext to an ExecuteContext."""
  return execute_context_impl.ExecuteContextImpl(
      canceller=context.canceller,

      geometry_service=context.geometry_service,

      logging_context=context.logging_context,
      motion_planner=context.motion_planner,
      object_world=context.object_world,
      resource_handles=resource_handles,
      context_id=context.context_id,
  )



def preview_to_predict_request(
    request: skill_interface.PreviewRequest[skill_interface.TParamsType],
) -> skill_interface.PredictRequest[skill_interface.TParamsType]:
  """Converts a PreviewRequest to a PredictRequest."""
  return skill_interface.PredictRequest(
      internal_data=request.internal_data,
      params=request.params,
  )


def preview_to_predict_context(
    context: skill_interface.PreviewContext,
) -> skill_interface.PredictContext:
  """Converts a PreviewContext to a PredictContext."""
  resource_handles = (
      context._resource_handles  # pylint: disable=protected-access
      if isinstance(context, preview_context_impl.PreviewContextImpl)
      else {}
  )

  return predict_context_impl.PredictContextImpl(

      geometry_service=context.geometry_service,

      motion_planner=context.motion_planner,
      object_world=context.object_world,
      resource_handles=resource_handles,
  )

File: intrinsic_sdk/intrinsic/skills/cc/skill_interface_utils.cc
// Copyright 2026 Intrinsic Innovation LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "intrinsic/skills/cc/skill_interface_utils.h"

#include <algorithm>
#include <memory>
#include <optional>
#include <string>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_format.h"
#include "absl/time/time.h"
#include "google/protobuf/message.h"
#include "intrinsic/skills/cc/equipment_pack.h"
#include "intrinsic/skills/cc/skill_interface.h"
#include "intrinsic/skills/internal/execute_context_view.h"
#include "intrinsic/skills/internal/predict_context_view.h"  
#include "intrinsic/skills/proto/prediction.pb.h"  
#include "intrinsic/skills/proto/skill_service.pb.h"
#include "intrinsic/util/proto_time.h"
#include "intrinsic/util/status/status_macros.h"
#include "intrinsic/world/proto/object_world_updates.pb.h"

namespace intrinsic {
namespace skills {

absl::StatusOr<std::unique_ptr<::google::protobuf::Message>> PreviewViaExecute(
    SkillExecuteInterface& skill, const PreviewRequest& request,
    PreviewContext& context) {
  EquipmentPack equipment;

  INTR_ASSIGN_OR_RETURN(ExecuteRequest execute_request,
                        PreviewToExecuteRequest(request));
  INTR_ASSIGN_OR_RETURN(ExecuteContextView execute_context,
                        PreviewToExecuteContext(context, equipment));

  return skill.Execute(execute_request, execute_context);
}


absl::Status PreviewViaPredict(SkillInterface& skill,
                               const PreviewRequest& request,
                               PreviewContext& context,
                               ::google::protobuf::Any* result_any) {
  INTR_ASSIGN_OR_RETURN(PredictRequest predict_request,
                        PreviewToPredictRequest(request));
  INTR_ASSIGN_OR_RETURN(PredictContextView predict_context,
                        PreviewToPredictContext(context));
  INTR_ASSIGN_OR_RETURN(intrinsic_proto::skills::PredictResult predict_result,
                        skill.Predict(predict_request, predict_context));

  if (predict_result.outcomes().empty()) {
    return absl::InternalError("Skill produced no predictions.");
  }

  const intrinsic_proto::skills::Prediction& most_likely_outcome =
      *std::max_element(predict_result.outcomes().begin(),
                        predict_result.outcomes().end(),
                        [](const intrinsic_proto::skills::Prediction& a,
                           const intrinsic_proto::skills::Prediction& b) {
                          return a.probability() < b.probability();
                        });

  // Record the world updates from the most likely outcome.
  absl::Time previous_start_time;
  absl::Duration previous_duration;
  for (int i = 0; i < most_likely_outcome.expected_states_size(); ++i) {
    intrinsic_proto::skills::TimedWorldUpdate expected_state =
        most_likely_outcome.expected_states(i);
    absl::Time start_time;
    if (expected_state.has_start_time()) {
      INTR_ASSIGN_OR_RETURN(start_time,
                            ToAbslTime(expected_state.start_time()));
    } else {
      start_time = previous_start_time + previous_duration;
    }
    INTR_ASSIGN_OR_RETURN(absl::Duration duration,
                          ToAbslDuration(expected_state.time_until_update()));

    absl::Duration elapsed = start_time - previous_start_time;
    if (elapsed < absl::ZeroDuration()) {
      return absl::InvalidArgumentError(absl::StrFormat(
          "Expected state start times must be monotonically increasing "
          "(state[%d] start time: %s, state[%d] start time: %s).",
          i, absl::FormatTime(start_time), i - 1,
          absl::FormatTime(previous_start_time)));
    }

    for (const intrinsic_proto::world::ObjectWorldUpdate& update :
         expected_state.world_updates().updates()) {
      INTR_RETURN_IF_ERROR(
          context.RecordWorldUpdate(update, elapsed, duration));

      // All updates in this expected state started at the same time, so all but
      // the first update should have `elapsed = 0`.
      elapsed = absl::ZeroDuration();
    }

    previous_start_time = start_time;
    previous_duration = duration;
  }

  if (result_any != nullptr) {
    *result_any = most_likely_outcome.result();
  }

  return absl::OkStatus();
}


absl::StatusOr<ExecuteRequest> PreviewToExecuteRequest(
    const PreviewRequest& request) {
  return ExecuteRequest(

      std::string(request.internal_data()),

      /*params=*/request.params_any(),
      /*param_defaults=*/std::nullopt);
}

absl::StatusOr<ExecuteContextView> PreviewToExecuteContext(
    PreviewContext& context, const EquipmentPack& equipment) {
  return ExecuteContextView(context.canceller(), equipment,
                            context.logging_context(), context.motion_planner(),
                            context.object_world()

                            ,
                            context.geometry_library()

                            ,
                            std::string(context.context_id()));
}


absl::StatusOr<PredictRequest> PreviewToPredictRequest(
    const PreviewRequest& request) {
  return PredictRequest(/*internal_data=*/std::string(request.internal_data()),
                        /*params=*/request.params_any(),
                        /*param_defaults=*/std::nullopt);
}

absl::StatusOr<PredictContextView> PreviewToPredictContext(
    PreviewContext& context) {
  PredictContextView predict_context(context.equipment(),
                                     context.motion_planner(),
                                     context.object_world()

                                     ,
                                     context.geometry_library()

  );

  return predict_context;
}


}  // namespace skills
}  // namespace intrinsic
